app/pages/lab-pages-editor/DataManager.jsx

```diff
@@ -57,19 +59,50 @@ function DataManager({
     fetchWorkflow();
   }, [workflowId]);
 
-  /*
-  Updates the workflow with new data.
-   */
-  function update(data) {
-    console.log('+++ TODO');
-  }
-
   // Wrap contextData in a memo so it doesn't re-create a new object on every render.
   // See https://react.dev/reference/react/useContext#optimizing-re-renders-when-passing-objects-and-functions
-  const contextData = useMemo(() => ({
-    workflow: apiData.workflow,
-    update
-  }), [apiData.workflow]);
+  const contextData = useMemo(() => {
+    console.log('+++ DataManager.useMemo');
+
+    /*
+    Updates the workflow with new data.
+    */
+    async function update(data) {
+      console.log('+++ DataManager.update()');
+      const wf = apiData.workflow;
+      if (!wf) return;
+
+      setApiData((prevState) => ({
+        ...prevState,
+        status: 'updating'
+      }));
+
+      const newWorkflow = await wf.update(data).save();
+
+      console.log('+++ are they the same? ', newWorkflow === wf);
+
+      // Note to self: hang on, will setting setApiData() cause the useMemo to update perpetually?
+      setApiData({
+        workflow: newWorkflow,
+        status: 'ready'
+      });
+
+      // Hmmm... this seems to be the only thing that's updating the context properly.
+      // Without it, it seems that setApiData() isn't changing the value
+      // (object reference) of`apiData.workflow`, which means the useMemo
+      // doesn't realise that oohhh wait I think I get it now.
+      // TODO: how to indicate that the value of the apiData.workflow object
+      // has changed when the change is local to the resource itself?
+      // Does the workflow object have an internal version tracker?
+      // @shaunanoordin 20230902
+      // setRandomFlagToPromptMemoUpdate(Math.floor(Math.random() * 10000));
+    }
+
+    return {
+      workflow: apiData.workflow,
+      update
+    };
+  }, [apiData.workflow, randomFlagToPromptMemoUpdate]);
```

Comment:
```suggestion
  }, [apiData.workflow, apiData.status]);
```

This fixes the editor for me when I run it locally.

Reply:
This might be a hack, but I think it’s cleaner than generating a random number on each state update.